pkg/controller/clusterpool/clusterpool_controller.go

```diff
@@ -0,0 +1,363 @@
+		}
+		builder.CloudBuilder = awsProvider
+
+	} else if clp.Spec.Platform.GCP != nil {
+		credsSecretName = clp.Spec.Platform.GCP.CredentialsSecretRef.Name
+		// Lookup the platform creds secret for this pool:
+		credsSecret := &corev1.Secret{}
+		if err := r.Client.Get(context.Background(), types.NamespacedName{Namespace: clp.Namespace, Name: credsSecretName}, credsSecret); err != nil {
+			logger.WithError(err).Error("error looking up credentials secret for pool in hive namespace")
+			return err
+		}
+
+		gcpSA := credsSecret.Data[constants.GCPCredentialsName]
+
+		projectID, err := gcpclient.ProjectID(gcpSA)
+		if err != nil {
+			return errors.Wrap(err, "error loading GCP project ID from service account json")
+		}
+
+		gcpProvider := &clusterresource.GCPCloudBuilder{
+			ServiceAccount: gcpSA,
+			ProjectID:      projectID,
+		}
+		builder.CloudBuilder = gcpProvider
+
+	} else if clp.Spec.Platform.Azure != nil {
+		credsSecretName = clp.Spec.Platform.Azure.CredentialsSecretRef.Name
+		// Lookup the platform creds secret for this pool:
+		credsSecret := &corev1.Secret{}
+		if err := r.Client.Get(context.Background(), types.NamespacedName{Namespace: clp.Namespace, Name: credsSecretName}, credsSecret); err != nil {
+			logger.WithError(err).Error("error looking up credentials secret for pool in hive namespace")
+			return err
+		}
+		azureSP := credsSecret.Data[constants.AzureCredentialsName]
+
+		azureProvider := &clusterresource.AzureCloudBuilder{
+			ServicePrincipal:            azureSP,
+			BaseDomainResourceGroupName: clp.Spec.Platform.Azure.BaseDomainResourceGroupName,
+		}
+		builder.CloudBuilder = azureProvider
+	}
+	// TODO: OpenStack, VMware, and Ovirt.
+
+	objs, err := builder.Build()
+	if err != nil {
+		return errors.Wrap(err, "error building resources")
+	}
+	for _, obj := range objs {
+		if err := r.Client.Create(context.Background(), obj); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *ReconcileClusterPool) obtainRandomNamespace(clp *hivev1.ClusterPool) (*corev1.Namespace, error) {
+	namespaceName := apihelpers.GetResourceName(clp.Name, utilrand.String(5))
+	ns := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: namespaceName,
+			Labels: map[string]string{
+				// Should never be removed.
+				constants.OriginClusterPoolNameLabel: clp.Name,
+			},
+		},
+	}
+	err := r.Create(context.Background(), ns)
+	return ns, err
+}
+
+func (r *ReconcileClusterPool) deleteExcessClusters(
+	clp *hivev1.ClusterPool,
+	poolCDs []*hivev1.ClusterDeployment,
+	deletionsNeeded int,
+	logger log.FieldLogger) error {
+
+	logger.Infof("too many clusters, searching for %d to delete", deletionsNeeded)
+	counter := 0
+	for _, cd := range poolCDs {
+		cdLog := logger.WithField("cluster", fmt.Sprintf("%s/%s", cd.Namespace, cd.Name))
+		if cd.DeletionTimestamp != nil {
+			cdLog.WithFields(log.Fields{"cdName": cd.Name, "cdNamespace": cd.Namespace}).Debug("cluster already deleting")
```

Comment:
Do we really need the "cdName" and "cdNamespace" fields when there is already the "cluster" field that includes the cd namespace and name?